# *M*ethylation-*I*nterruption-*E*valuated Local Discordance (MIELD)
Besides global hypomethylation and focal hypermethylation, local DNA methylation discordance emerges as a new feature of tumor methylome. We quantified methylation discordance based on the methylation interruption events captured by bisulfite sequencing. The methylation interruption events are represented by the unmethylated CpGs in partially methylated reads (slashed circles in Fig 1a). So, bisulfite sequencing reads are dissected into three categories of fragments, i.e. methylated fragments (consecutive solid circles in Fig 1a), unmethylated fragments (consecutive blank circles), and methylation interrupted fragments (consecutive slashed circles). Thus, DNA Methylation-interruption-evaluated local discordance (MIELD) score of a particular genomic region is measured by the following equation. 
MIELD=(∑_(i=1)^I▒ω_i )/(∑_(i=1)^I▒ω_i +∑_(m=1)^M▒ω_m +∑_(u=1)^U▒ω_u )
M, U and I represent the numbers of methylated fragments, unmethylated fragments and methylation interrupted fragments, respectively.  ω_m, ω_u and ω_i are the weights for each fragment. Optional weights can be the CpG counts of each fragment or 1.
